### ISSUE TITLE
feat(strategies): add verifiability-routing-heuristic — assess task verifiability before choosing prompting vs fine-tuning

### ISSUE THREAD

**Issue by visionik** [deftai/directive#781]
## Summary

Encode a named strategy heuristic: before designing an LLM feature, assess whether the task is in a verifiable domain. Tasks outside verified RL training circuits require different approaches than prompting alone.

## Source

Karpathy AI Ascent talk (agentic engineering). Karpathy's verifiability framework: LLMs excel in verifiable domains (math, code) because frontier labs apply RL with verification rewards to those domains. Tasks outside these "circuits" — where the model wasn't trained with RL rewards — produce jagged, unreliable results regardless of prompting quality.

Example: state-of-the-art models can refactor a 100,000-line codebase but will tell you to walk to a car wash 50 metres away. This jaggedness is structural, not a prompting failure.

## Why This Belongs in Directive

Directive guides projects to use LLMs for automation. Without this heuristic, teams apply prompting uniformly — and blame prompt quality when models struggle in non-verifiable domains. The correct diagnosis is structural: if the task is outside verifiable circuits, prompting improvement has diminishing returns. Fine-tuning, RAG, or alternative approaches should be evaluated instead.

Heuristic to encode:
1. Is the task verifiable (can a function or test confirm correctness)?
2. Is it a domain labs have focused RL training on?
3. If no to either: do not rely on prompting alone — evaluate fine-tuning or decomposition.

## Files Affected

- `strategies/` — new heuristic or section in an existing strategy file
- `patterns/` — possibly a decision-tree pattern for LLM feature design

## Confidence

Medium — well-grounded in Karpathy's writing but requires careful operationalisation to avoid over-engineering.

## Related

Related to #691 (domain-finetuned LLMs), #688 (agent memory decision framework), #606 (LLM evals as a testing tier).

**Comment by visionik** [comment]
## Autobrowse: regime mismatch + cheapest-primitive-first rule

Source: Kyle Jeong, "Autobrowse" (956 bookmarks, 201K impressions)
https://x.com/kylejeong/status/2052103973377867913

The article contains a costly concrete example of the verifiability mismatch this issue aims to address, plus a routing rule.

**The regime mismatch failure — 4 on a static HTML table:**

> "We threw Autobrowse at it anyway, because the framing of 'let the agent figure it out' is seductive. Four iterations and ~4 later, the loop still hadn't returned all 167 rows in a single output."

The task: a 167-row static HTML catalog. No JavaScript, no auth, no anti-bot. The agent's per-turn output cap kept truncating reasoning on a problem that didn't reward cleverness. Once the regime was recognized, 200 lines of deterministic Python with BeautifulSoup solved it instantly at zero inference cost.

The lesson was written into the graduated skill: "If the data is in the HTML, write the parser first."

**The agency-level routing rule:**

Autobrowse encodes a concrete primitive ordering:

``````````````````````````````````\xample.gov/programs"
# If the data comes back cleanly i# If the data comes back cleanly i# If the data comes back cleanly i# If the data comes back cleanly i#
```

The principle: The principle: The principle: The principle: ore escalating to a The principle: The principle: The principle: The principle: ore escalating to a The principle: The principle: The principle: The principle: ore escalating to a The princip deterministic: script, not LLM


he principle: The principle: The principle: The principle: ore escalating to a The principle: The principle: The principle: The principle: ore escalating to a The principle: The principle: The principle: The principle: ore escalating to a The princip deterministic: script, not LLM
that doesn't reward cleverness (static HTML parsing) wastes money on high-agency approaches regardless of how well the agent executes.

The routing question is not just "can I verify the output?" but "does this task rThe routing question is not just "can I verify the output?" but "does this task rThe rely and go straight to code.

**Comment by visionik** [comment]
## Autobrowse: regime mismatch + cheapest-primitive-first rule

Source: Kyle Jeong, "Autobrowse" (956 bookmarks, 201K impressions)
https://x.com/kylejeong/status/2052103973377867913

The article contains a costly concrete example of the verifiability mismatch this issue aims to address, plus a routing rule.

**The regime mismatch failure — 4 on a static HTML table:**

> "We threw Autobrowse at it anyway, because the framing of 'let the agent figure it out' is seductive. Four iterations and ~4 later, the loop still hadn't returned all 167 rows in a single output."

The task: a 167-row static HTML catalog. No JavaScript, no auth, no anti-bot. The agent's per-turn output cap kept truncating reasoning on a problem that didn't reward cleverness. Once the regime was recognized, 200 lines of deterministic Python with BeautifulSoup solved it instantly at zero inference cost.

The lesson was written into the graduated skill: "If the data is in the HTML, write the parser first."

**The agency-level routing rule:**

Autobrowse encodes a concrete primitive ordering:

```bash
# Step 1: probe with fetch first.
browse fetch "https://example.gov/programs"
# If the data comes back cleanly i# If the data comes back cleanly i# If the data comes back cleanly i# If the data comes back cleanly i# If the data comes bae: # If the data comes back cleanly i# If the data comes back clo a more expensive one**.

Applied generally:
1. Static fetch → parse → done
2. If gated/dynamic: browser session
3. If novel/complex: full Autobrowse iteration loop
4. If deterministic: script, not LLM

**How this sharpens #781 (verifiability-routing-heuristic):**

The issue asks "assess task verifiability before choosing prompting vs. fine-tuning." Autobrowse adds a more operational framing: assess whether the task **rewards cleverness** before choosing the agency level. A task that doesn't reward cleverness (static HTML parsing) wastes money on high-agency approaches regardless of how well the agent executes.

The routing question is not just "can I verify the output?" but "does this task require reasoning, or just execution?" Deterministic tasks should skip the LLM tier entirely and go straight to code.

**Comment by visionik** [comment]
## Autobrowse: regime mismatch + cheapest-primitive-first rule

Source: Kyle Jeong, "Autobrowse" (956 bookmarks, 201K impressions)
https://x.com/kylejeong/status/2052103973377867913

The article contains a costly concrete example of the verifiability mismatch this issue aims to address, plus a routing rule.

**The regime mismatch failure — $24 on a static HTML table:**

> "We threw Autobrowse at it anyway, because the framing of 'let the agent figure it out' is seductive. Four iterations and ~$24 later, the loop still hadn't returned all 167 rows in a single output."

The task: a 167-row static HTML catalog. No JavaScript, no auth, no anti-bot. The agent's per-turn output cap kept truncating reasoning on a problem that didn't reward cleverness. Once the regime was recognized, 200 lines of deterministic Python with BeautifulSoup solved it instantly at zero inference cost.

The lesson was written into the graduated skill: "If the data is in the HTML, write the parser first."

**The agency-level routing rule:***The agency-level routing rule:***The agency-level routing r
# Step 1: probe with fetch first.
browse fetch "https://example.gov/pbrowse fetch "https://example.gov/pbrowsy in the resbrowse fetch "https://example.gov/pbrowse fetch "https://example.gov/pbrowsy in the resbrowse fetch "httprinciple: **reach for the cheapest viable primitive before escalating to a more expensibrowse fetch "https://example.gov/pbrowse fetch "https://example.gov/pbrowsy in the resbrowse fetch "https://example.gov/pbrowse fetch "https://example.gov/pbrowsy in the resbrowse fetch "httprinciplhisbrowse fetch "https://example.gov/pbrowse fetch "https://example.gov/pbrowsy in the resbrowse fetch "https://example.gov/pbrowse fetch "https://example.gov/pbrowsy in the resbrowse fetch "httprinciple: **reach for the cheapest viable primitive before escalating to a more expensibrowse fetch "https://examptic HTML parsing) wastes money on high-agency approaches regardless of how well the agent executes.

The rouThe rouThe rouThe rouThe rouThe rouThe rouThe rouThe rouThe rouThe rouThe rouThe rouThe rou or just execution?" Deterministic tasks should skip the LLM tier entirely and go straight to code.

**Comment by visionik** [comment]
## Autobrowse: regime mismatch + cheapest-primitive-first rule

Source: Kyle Jeong, "Autobrowse" (956 bookmarks, 201K impressions)
https://x.com/kylejeong/status/2052103973377867913

The article contains a costly concrete example of the verifiability mismatch this issue aims to address, plus a routing rule.

**The regime mismatch failure —  on a static HTML table:**

> "We threw Autobrowse at it anyway, because the framing of 'let the agent figure it out' is seductive. Four iterations and ~ later, the loop still hadn't returned all 167 rows in a single output."

The task: a 167-row static HTML catalog. No JavaScript, no auth, no anti-bot. The agent's per-turn output cap kept truncating reasoning on a problem that didn't reward cleverness. Once the regime was recognized, 200 lines of deterministic Python with BeautifulSoup solved it instantly at zero inference cost.

The lesson was written into the graduated skill: "If the data is in the HTML, write the parser first."

**The agency-level routing rule:**

Autobrowse encodes a concrete primitive ordering:

https://github.com/deftai/directive/issues/781#issuecomment-4406856802
https://github.com/deftai/directive/issues/781#issuecomment-4406859579
https://github.com/deftai/directive/issues/781#issuecomment-4406862436
Exit status: 0

The principle: **reach for the cheapest viable primitive before escalating to a more expensive one**.

Applied generally:
1. Static fetch → parse → done
2. If gated/dynamic: browser session
3. If novel/complex: full Autobrowse iteration loop
4. If deterministic: script, not LLM

**How this sharpens #781 (verifiability-routing-heuristic):**

The issue asks "assess task verifiability before choosing prompting vs. fine-tuning." Autobrowse adds a more operational framing: assess whether the task **rewards cleverness** before choosing the agency level. A task that doesn't reward cleverness (static HTML parsing) wastes money on high-agency approaches regardless of how well the agent executes.

The routing question is not just "can I verify the output?" but "does this task require reasoning, or just execution?" Deterministic tasks should skip the LLM tier entirely and go straight to code.

**Comment by visionik** [comment]
## Rule 5 operationalizes verifiability-routing at the code level: don't use the model for deterministic work

Source: @Mnilax, "Karpathy's 4 CLAUDE.md rules" (3103 bookmarks)
https://x.com/Mnilax/status/2053116311132155938

The article's Rule 5 makes the verifiability-routing decision operational at the code level:

> "Use Claude for: classification, drafting, summarization, extraction from unstructured text.
> Do NOT use Claude for: routing, retries, status-code handling, deterministic transforms.
> If a status code already answers the question, plain code answers the question."

**The failure mode it addresses:**

> "Code that called Claude to 'decide if we should retry on 503' worked beautifully for two weeks, then started flaking because the model started reading the request body as context for the decision. The retry policy was random because the prompt was random."

**How this sharpens this issue (#781):**

#781 asks: "assess task verifiability before choosing prompting vs. fine-tuning." Rule 5 applies the same logic inside code, not just to tasks: don't call the model for decisions that have deterministic answers. The model becomes a random variable in a path that needs to be a fixed function.

The concise routing test:
- Can this be answered by code alone (status code, regex, lookup, calculation)? → Use code
- Does this require language understanding (classification, extraction, drafting)? → Use model
- Is this potentially either? → Default to code; escalate to model only when code gives up

**Proposed addition to this issue:**

Add a code-level variant of the verifiability-routing heuristic: within agent-generated code, model calls should be reserved for judgment tasks. Routing logic, retry policies, format detection, numeric thresholds — these should be deterministic code, not model calls. This is the regime mismatch principle (Autobrowse, #781 earlier comment) applied at the function level.
